### PR TITLE
Fix zoom 100/200 bounds also in second window

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
As done in #8870 the 100% & 200% bounds while zooming in via mouse scroll should be respected in a second window.

Can't test myself, mayby @MStraeten ?

### CODE REVIEW

**Comment by MStraeten** [comment]
The issue seems to be more complex on OSX (constellation: retina 16" display + external 27" WQHD screen)

without this pr:
darktable main - retina, preview - retina: both full size at 100% compared to apple preview/gimp
darktable main - retina, preview - wqhd: both full size
darktable main - wqhd, preview - retina: main half size, preview full size
darktable main - wqhd, preview - wqhd: main full size, preview full size

with this pr: 
darktable main - retina, preview - retina: both half size at 100% compared to apple preview/gimp
darktable main - retina, preview - wqhd: both half size 
darktable main - wqhd, preview - retina: main half size, preview full size
darktable main - wqhd, preview - wqhd: main half size, preview full size

I prefer the old way since there is a way to get full size 100% in the main edit window.

**Comment by jenshannoschwalm** [comment]
I think what you describe is related to the way the ppd for a window is "detected". For osx this is sort of constant as discussed in #7910. 

https://github.com/darktable-org/darktable/blob/688c91aa24a9061268a36da3719c0fe66bfa6634/src/gui/gtk.c#L1568-L1578

Don't know how or if this can be fixed in a proper way, at least for me that is the culprit and out of this pr. Could you test whether the special treatment for osx is still necessary? Maybe @parafin has an opinion here again?

**Comment by parafin** [comment]
What is 100% on High DPI (Retina) screen? 1:1 real screen pixel, or upscaled?
@MStraeten, full size means upscaled, and half-size is 1:1?

**Comment by MStraeten** [comment]
#7910 was fixed via #8870 

Maybe I don't get the issue this pr wants to solve.
- consistency of size to other image displaying tools?
- consistency of the size of main and preview window? 

both is stuff i don't bother about as long as i'm able to zoom in or out to the level i need - and thats pretty fine with darktable ;) 
a further more important topic:
Moving a darktable window from internal to external screen within a session doesn't change the displayed area of the image (in difference to krita - there moving the window results in a rescale and so different displayed area) - thats more important for me then 100% in relation to display resolution, and thats fine with darktable

@parafin "100%" is for me darktable zoomed in until the zoom indicator shows 100%. If also "original size" in apple preview or (1:1) gimp is displayed in the same size this is a good indicator for me that darktable handles "100%" in a proper way.
At least for my system configuration (retina display and external wqhd) the current master behavior (including #8870) is fine and this pr introduce a darktable specific interpretation of "100%" thats not identical to "original size" or "1:1" in apple preview or gimp. 

tl;dr: no need to deal with this in my point of view

**Comment by jenshannoschwalm** [comment]
This pr fixes the same issue for the second window as @aurelienpierre mentioned for main. The zooming in via mouse scrolling didn't stop at 100% but at 200%. Was commit `Fix scaling using mouse scrolls for ppd` in #8870

**Comment by parafin** [comment]
@MStraeten, my question was how large is 100%. So say you have a Retina display with native resolution 3072x1920 and you open a photo with exactly the same resolution. In 100% zoom (original size, 1:1, whatever) is/should it fit into the whole screen, or will you see only 1/4th of the photo, due to UI scaling (1 dot being 2x2 pixels)? I think there was a related issue about that, will try to look for it.

**Comment by MStraeten** [comment]
100% in darktable is 100% on the retina screen - same size
screenshot (left) in darktable (right):
image

**Comment by jenshannoschwalm** [comment]
100% is understood always to be true-pixel-wise. That's at least what @aurelienpierre required. So on a retina screen the image should fill whole screen. Agreed?

**Comment by parafin** [comment]
OK, and I assume screenshot has native (full) screen resolution.

I can see the problem then - we need to know screen PPD to know what 100% is. Since on macOS currently we detect PPD only for main window, it can be wrong for preview window. So I expect it to work like this:

1). Main HighDPI, preview normal - preview window at 100% will be smaller (zoomed out to 50%).
2). Main normal, preview HighDPI - preview window at 100% will be bigger (zoomed in to 200%).
3). Main normal, preview normal or main HighDPI, preview HighDPI - everything should work fine.

For some reason it doesn't match what @MStraeten reported.

**Comment by parafin** [comment]
The problem with 100% meaning 1:1 native screen pixels is that it's much smaller zoom in physical measurement when comparing HighDPI to normal screen. So say when user switches the display from normal to HighDPI, he will probably would want to go to twice the zoom level he did before. IMHO we should match what other applications are doing here, and I'm not sure what it is.

**Comment by MStraeten** [comment]
my 100% comparison was made with current master, updated today.
The investigation from yesterday were made with a build made 3 days ago (darktable 3.5.0+2158~g066428e9df-dirty so maybe without having the latest #8870 commits)
doing the 100% comparison with that build it's about double the size. (main and preview as well)

Current master is:
darktable main on integrated display: main 1:1 and preview double size both on integrated display as well as external screen
darktable on external screen: double size as on retina screen and preview also double size on integrated display as well as external screen

**Comment by jenshannoschwalm** [comment]
> we need to know screen PPD to know what 100% is ...

I would not say "screen PPD" but "widget PPD" to make sure this could/can be updated  while moving a window from one screen to another. And that's a remaining issue as i see it.

> The problem with 100% meaning 1:1 native screen pixels is that it's much smaller zoom in physical measurement when comparing HighDPI to normal screen

Yes, and i personally dislike this but it seems what other people are expecting. 100% seems to be understood as: 1pixel in image represents one physical dot on display.

**Comment by TurboGit** [comment]
> 100% seems to be understood as: 1pixel in image represents one physical dot on display.

Which is safe as it does not need any over sampling. 1:1 really means that a pixel is not changed in anyway and to me this is important to evaluate the sharpness, no?

**Comment by jenshannoschwalm** [comment]
> to me this is important to evaluate the sharpness, no?

I think so too. But remember, this pr should not (and does not) change the way the image is rendered. It just makes the zooming-by-mouse-scroll stopping at the 100% and 200% levels for the second window if the control key is not pressed.

So a 100%-scaled image in dr has a different size on low- vs high-res displays if both have the same physical dimensions. This is different from what we expect for image viewers or text processors...

I simulated the situation by declaring my 2560x1440 screen1 as ppd==2 and the TV as ppd==1 and everything was fine for me - on linux the ppd is calculated from the widget.

For osx there is the `dt_osx_get_ppd();` function returning the same ppd for both screens and as we have just one ppd defined via preferences this doesn't help here too.

**Comment by TurboGit** [comment]
@jenshannoschwalm : so should we go ahead with this ? I cannot test myself, the code looks ok to me otherwise.

**Comment by jenshannoschwalm** [comment]
> so should we go ahead with this ?

I really think so.

**Comment by TurboGit** [comment]
> > so should we go ahead with this ?
> 
> I really think so.

Ok, so please fix the conflict on darkroom.c and I'll do a pass on this tomorrow. Thanks.

**Review by TurboGit — APPROVED** [review]
Thanks !
